### PR TITLE
[SPARK-9116] [SQL] [PYSPARK] support Python only UDT in __main__

### PR DESCRIPTION
**Pull request by davies**
Also we could create a Python UDT without having a Scala one, it's important for Python users.

cc @mengxr @JoshRosen

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala [hunk 1]
```diff
@@ -82,3 +82,22 @@ abstract class UserDefinedType[UserType] extends DataType with Serializable {
...
+
+/**
+ * ::DeveloperApi::
+ * The user defined type in Python.
+ *
+ * Note: This can only be accessed via Python UDF, or accessed as serialized object.
+ */
+class PythonUserDefinedType(val sqlType: DataType, pyClass: String) extends UserDefinedType[Any] {
+  override def pyUDT = pyClass
+  def serialize(obj: Any): Any = obj
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala]: are these interfaces designed in UserDefinedType? If yes, can you add override here? it caches a lot of potential problems when we do refactoring down the line.

File: python/pyspark/mllib/linalg.py [hunk 2]
```diff
@@ -778,6 +778,8 @@ def dense(elements):
         >>> Vectors.dense([1, 2, 3])
         DenseVector([1.0, 2.0, 3.0])
         """
+        if isinstance(elements, (int, float)):
```
Comment (JoshRosen) [inline python/pyspark/mllib/linalg.py]: This conflicts with / overlaps with the changes in #7476, so I guess this patch is blocking until that other one gets reviewed?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/pythonUDFs.scala [hunk 3]
```diff
@@ -271,7 +271,6 @@ object EvaluatePython {
           pickler.save(row.values(i))
           i += 1
         }
-        row.values.foreach(pickler.save)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/pythonUDFs.scala]: Wow, this looks like a bug that I overlooked while reviewing #7301; nice catch.

File: python/pyspark/mllib/linalg.py (hunk 2 above)
Comment (davies) [inline python/pyspark/mllib/linalg.py]: yes

File: python/pyspark/sql/types.py [hunk 4]
```diff
@@ -641,6 +642,50 @@ def __eq__(self, other):
...
+        return '__main__'
+
+    def serialize(self, obj):
+        return [obj.x, obj.y]
+
+    def deserialize(self, datum):
+        return Point(datum[0], datum[1])
+
+    def __eq__(self, other):
+        return True
```
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: Why not just use the default `__eq__`?

File: python/pyspark/sql/types.py [hunk 5]
```diff
@@ -580,7 +581,7 @@ def scalaUDT(cls):
         """
         The class name of the paired Scala UDT.
```
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: Should we update the comment to clarify that we return an empty string if there is no corresponding Scala UDT?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 6]
```diff
@@ -144,10 +144,14 @@ object DataType {
...
-    ("pyClass", _),
-    ("sqlType", _),
+    ("pyClass", pyClass: JValue),  // pyClass could be null, cannot match with JString()
+    ("sqlType", v: JValue),
     ("type", JString("udt"))) =>
-      Utils.classForName(udtClass).newInstance().asInstanceOf[UserDefinedType[_]]
+      if (udtClass.length > 0) {
+        Utils.classForName(udtClass).newInstance().asInstanceOf[UserDefinedType[_]]
+      } else {
+        new PythonUserDefinedType(parseDataType(v), pyClass.asInstanceOf[JString].values)
```
Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: I guess we only need UDT support in Python + JVM for now, but I wonder whether assuming that an empty string implies Python is the right long-term approach if we ever decide to support, say, R UDTs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 7]
```diff
@@ -144,10 +144,14 @@ object DataType {
 
     case JSortedObject(
     ("class", JString(udtClass)),
-    ("pyClass", _),
-    ("sqlType", _),
+    ("pyClass", pyClass: JValue),  // pyClass could be null, cannot match with JString()
+    ("sqlType", v: JValue),
```
Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: I thought that Python UDTs had to implement `sqlType`?

File: python/pyspark/sql/types.py [hunk 8]
```diff
@@ -641,6 +642,50 @@ def __eq__(self, other):
         return type(self) == type(other)
 
 
+class PointUDT(UserDefinedType):
```
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: This class is only used in tests, right?  If so, I think we should move it to a tests file.

File: python/pyspark/sql/types.py [hunk 9]
```diff
@@ -656,7 +701,7 @@ def _parse_datatype_json_string(json_string):
     ...     assert datatype == pickled
     ...     scala_datatype = sqlContext._ssql_ctx.parseDataType(datatype.json())
     ...     python_datatype = _parse_datatype_json_string(scala_datatype.json())
-    ...     assert datatype == python_datatype
+    ...     assert datatype == python_datatype, str(datatype) + str(python_datatype)
```
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: What role is the comma playing here?  What does this actually end up asserting?

File: python/pyspark/sql/types.py (hunk 8 above)
Comment (JoshRosen) [inline python/pyspark/sql/types.py]: It looks like the original `ExamplePointUDT` is still in the tests file, too.  Why not just use that?

File: python/pyspark/cloudpickle.py [hunk 10]
```diff
@@ -350,7 +350,26 @@ def save_global(self, obj, name=None, pack=struct.pack):
...
+            self.save(_load_class)
+            self.save_reduce(typ, (obj.__name__, obj.__bases__, {"__doc__": obj.__doc__}), obj=obj)
+            d.pop('__doc__', None)
+            # handle property and staticmethod
+            dd = {}
+            for k, v in d.items():
+                if isinstance(v, property):
+                    k = ('property', k)
+                    v = (v.fget, v.fset, v.fdel, v.__doc__)
+                elif isinstance(v, staticmethod) and hasattr(v, '__func__'):
```
Comment (JoshRosen) [inline python/pyspark/cloudpickle.py]: Looks like PyLint is complaining about a possibly undefined loop variable `v` at this line.  If this isn't a legitimate error, then we can just add a comment to bypass that warning here.

Comment (davies) [inline python/pyspark/cloudpickle.py]: I think it's annoying to let PyLint report Warning as Error, should we only fail on real errors?

Comment (JoshRosen) [inline python/pyspark/cloudpickle.py]: Good point; let's see if we can update the configuration to do that.

Comment (davies) [inline python/pyspark/cloudpickle.py]: I tried with putting `-E` to pylint, it seems even worse, lots of false-negtive errors.

**Comment by davies** [comment]
@mengxr Could you help to review this?

**Comment by JoshRosen** [comment]
Ping @mengxr, I think we need to get this in for 1.5.0 since it contains a bugfix for another patch.

**Comment by mengxr** [comment]
I'll make a pass tomorrow.

File: python/pyspark/cloudpickle.py [hunk 11]
```diff
@@ -708,6 +727,20 @@ def _make_skel_func(code, closures, base_globals = None):
                               None, None, closure)
 
 
+def _load_class(cls, d):
```
Comment (mengxr) [inline python/pyspark/cloudpickle.py]: missing doc

File: python/pyspark/sql/context.py [hunk 12]
```diff
@@ -277,6 +277,59 @@ def applySchema(self, rdd, schema):
 
         return self.createDataFrame(rdd, schema)
 
+    def _createFromRDD(self, rdd, schema, samplingRatio):
```
Comment (mengxr) [inline python/pyspark/sql/context.py]: missing doc, especially the return values

File: python/pyspark/sql/context.py [hunk 13]
```diff
@@ -277,6 +277,59 @@ def applySchema(self, rdd, schema):
...
+            schema = struct
+
+        elif isinstance(schema, StructType):
+            # take the first few rows to verify schema
+            rows = rdd.take(10)
+            for row in rows:
+                _verify_type(row, schema)
+
+        else:
+            raise TypeError("schema should be StructType or list or None")
```
Comment (mengxr) [inline python/pyspark/sql/context.py]: include `schema` in the error message

File: python/pyspark/sql/context.py [hunk 14]
```diff
@@ -277,6 +277,59 @@ def applySchema(self, rdd, schema):
...
+
+        # convert python objects to sql data
+        rdd = rdd.map(schema.toInternal)
+        return rdd, schema
+
+    def _createFromLocal(self, data, schema):
```
Comment (mengxr) [inline python/pyspark/sql/context.py]: missing doc

File: python/pyspark/sql/context.py [hunk 15]
```diff
@@ -277,6 +277,59 @@ def applySchema(self, rdd, schema):
...
+                    struct.fields[i].name = name
+                    struct.names[i] = name
+            schema = struct
+
+        elif isinstance(schema, StructType):
+            for row in data:
+                _verify_type(row, schema)
+
+        else:
+            raise TypeError("schema should be StructType or list or None")
```
Comment (mengxr) [inline python/pyspark/sql/context.py]: ditto, include input `schema` in the error message

File: python/pyspark/sql/tests.py [hunk 16]
```diff
@@ -75,11 +75,7 @@ def sqlType(self):
 
     @classmethod
     def module(cls):
-        return 'pyspark.tests'
-
-    @classmethod
-    def scalaUDT(cls):
-        return 'org.apache.spark.sql.test.ExamplePointUDT'
```
Comment (mengxr) [inline python/pyspark/sql/tests.py]: Do we also want to keep the test for UDTs that works for both Scala and Python?

**Comment by mengxr** [comment]
Made one pass. My main comment is to keep the unit test for UDTs that work in both Python and Scala/Java.

File: python/pyspark/sql/types.py (hunk 9 above)
Comment (davies) [inline python/pyspark/sql/types.py]: It's kind of error message.

**Comment by davies** [comment]
@mengxr Had kept the unit tests for Python and Scala UDT, please take another round review.

**Comment by mengxr** [comment]
LGTM

**Comment by davies** [comment]
Merged into master.

**Pull request closed without merging**